web/app/services/application.js

```diff
@@ -0,0 +1,14 @@
+/* Service for storing the main state of the application */
+
+class ApplicationService {
```

Comment:
Main application state should be hold by the redux store.
I suggest creating a `store.js` service that holds the application state (the redux code)
What you have in this file should probably more be in the helpers folder? Happy to call it `helpers/application`.

Reply:
That keep me thinking, because you are right, it's not a service, because it doesn't contain any data,  but making a helper to say if it's in localhost or production doesn't get much sense, helpers usually are for doing more specific jobs that doesn't fit much in the application, like an specific algorithm.
But beyond not needing any data to know if it's production, it's still part of a state of the application, because it changes the behaviour of it.

Talking about the state, i readed Redux documentation, and i don't see it as a good option this project. Having all the data in a God Object(like Redux does), makes all the views (and possible components) tidy to a single object, makes division of responsibility too complicated over the state, complicating to breaking down into simpler model interfaces, and the need of being constantly dehidrating and rehidrating the models on every change. And also every change fires every observer, because there is only one state. I really don't like it, Using a service/provider  pattern with Observables would make it easy to divide the logic.

What do you think about this ?

Reply:
Hi @BrianCraig,

Let me start by replying you on the redux question. Redux is pretty low-level, so all the things you mention can be solved.
For example, to avoid firing every observer, see https://github.com/tmrowco/electricitymap/blob/master/web/app/main.js#L1494. It watches for specific changes.
Second, components should never be tied to the store object. Instead, they can observe changes of sub-portions of the state.
Actually all of this is pretty well defined in React, where `mapStateToProps` projects the global state to a subset that corresponds to properties exclusively required by the component, thereby making the interface cleaner. `propTypes` can help having type safety on that interface.

The important thing we need to have regarding state management is make sure it is a one-way-flow. Redux handles this by requiring state changes to be sent using `dispatch`, and processing state changes using `reducers`.

I'm happy to refactor that part of the application as I have a bit of experience in this, but it would depend on which framework we use, so maybe we should wait a bit with that.

For now, what about we try to keep all application state in one file? Right now it means redux store object + the methods you've put in application state. Ideally the method should be called upon construction, and should populate the state object so they can be observed for changes. Not a big deal for now though.

Reply:
Oh well, i was pretty ignorant (still i am), i think you're con the right track.

I think i need to learn Redux, and some React too!

I expect to commit the merge of the state and application in to a redux by today, after learning some redux.

Cheers!